### PR TITLE
Remove parts of AOTDedupWrapper logic since we have dynamo dedup

### PR DESCRIPTION
**Pull request by jamesjwu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127306

Since dynamo already handles deduping arguments, we are choosing to decrease the complexity of AOTDedupWrapper by removing the parts of it that save state or add guards. Specifically, we can raise a runtime error if Strategy 1 is not sufficient for deduping, which will never happen in cases with dynamo. 

We still use the "leafify tensors" strategy of dedup to handle any cases in non-strict export where there are duplicate inputs that aren't mutated. This is tested by `test_aot_export_input_dupes_banned`. For any cases where strategy 1 is not sufficient AND dynamo isn't used to dedup args, we throw an exception. 

I've added some tests in test_aot_autograd, which may be somewhat duplicate of existing tests, but wanted to copy them over from tests that we now delete. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -2836,119 +2836,6 @@ def compiler(fx_g, _):
...
-    def test_invalid_dupe(self, counter):
-        self._test_invalid_dupe(counter, fake=False)
-
-    # See Note: Dynamo recompilation guarding invalid grad for why this test exists
-    @patch("torch._functorch.aot_autograd.AOT_COUNTER", new_callable=itertools.count)
-    @patch("torch._functorch.config.debug_assert", True)
-    def test_invalid_dupe_fake(self, counter):
-        self._test_invalid_dupe(counter, fake=True)
-
-    def _test_invalid_dupe(self, counter, fake):
```
Comment (jamesjwu) [inline test/functorch/test_aotdispatch.py:2907]: test_arg_dupe_via_dynamo_recompiles is the equivalent of this test

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 2]
```diff
@@ -673,257 +669,72 @@ def inner_fn(args: List[Any]):
...
+    Encountered duplicated inputs that are mutated in the graph you are trying to export.
+    This functionality is currently not supported. If needed, please file a github issue.
 
-        debugged_compiled_fn._boxed_call = True  # type: ignore[attr-defined]
+    fw_metadata={str(fw_metadata)}
+        """
+        )
 
-        return debugged_compiled_fn
+    raise RuntimeError(
```
Comment (jamesjwu) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:726]: This is the new error we throw if autograd encounters duplicate arguments that it doesn't know how to handle. Theoretically, this never runs when dynamo is involved.

**Comment by ezyang** [comment]
Do you want review now?

**Comment by jamesjwu** [comment]
Not yet, there's some issues with the PR and I need to decide exactly how much I want to try to remove

**Pull request closed without merging**